### PR TITLE
Ordering of actions

### PR DESCRIPTION
**Pull request by fzipi**
This PR hadles reordering of actions.

The final order that I used was:
```
id
phase
disruptive-action
status
capture
t:xxx
log
nolog
auditlog
noauditlog
msg
logdata
tag
sanitiseArg
sanitiseRequestHeader
sanitiseMatched
sanitiseMatchedBytes
ctl
setenv
*rev
*ver
*severity
setvar
expirevar
chain
skip
skipAfter
```

I specifically didn't change the TX:PARANOIA_LEVEL one liners. What do you think? Should the one-liners be changed also?

### CODE REVIEW

**Comment by dune73** [comment]
That's an impressive piece of work. Thank you @fzipi.

I did a first visual inspection focusing on "is anything missing" of your PR. I spotted bugs in the following rules:

```
920160
920450
933180
941260
943110
```

I presume, we decided to allow for multiple transformations to stay on the same line. Which actually makes some sense if we think of it as a transformation pipeline.

As far as PL helper rules go, I suggest to address them in a separate small PR after this one is merged.

This one needs more eyes though. The faster it is merged the better.

**Comment by fzipi** [comment]
I fixed 920160 933180 941260 943110, but didn't spotted anything on 920450. I've read it a dozen times, but didn't spotted the problem. Please enlighten me!

I concur that the faster is merged the better.

**Comment by dune73** [comment]
Sorry, I think I highlighted this before I realised we want to allow transformations on a single line.

I confirm: 920450 is OK.

Thank you for fixing the other ones.

Do you think it is possible to write a travis script that can check the correct order of ids in (chained)! SecRules for the future?

**Comment by dune73** [comment]
I'm getting two syntax errors now for the rule set. Did not spot them before.

**Comment by fzipi** [comment]
Yeah, definitely. Started my own container to check for it, and now apache says syntax is ok. 

There should be a travis check for this...

**Comment by fzipi** [comment]
For checking the correct order in travis, I think something could be made. Let me think about it.

**Comment by dune73** [comment]
Syntax OK now. Thank you.

I have done tests with over 100K requests on this PR and the behaviour (-> alerts triggered) is identical to v3.1/dev. This is no guarantee for being bug-free, but AFAICS there is no change in behaviour.

So, as far as I am concerned, this PR is ready to be merged.

If somebody sees a reason to hold on until other PRs have been merged, then please speak up.

File: rules/REQUEST-910-IP-REPUTATION.conf [hunk 1]
```diff
@@ -76,8 +76,8 @@ SecRule TX:HIGH_RISK_COUNTRY_CODES "!@rx ^$" \
             setvar:tx.anomaly_score=+%{tx.critical_anomaly_score},\
             setvar:tx.%{rule.id}-AUTOMATION/MALICIOUS-%{matched_var_name}=%{matched_var},\
             setvar:ip.reput_block_flag=1,\
-            expirevar:ip.reput_block_flag=%{tx.reput_block_duration},\
-            setvar:'ip.reput_block_reason=%{rule.msg}'"
+            setvar:'ip.reput_block_reason=%{rule.msg}'\
+	    expirevar:ip.reput_block_flag=%{tx.reput_block_duration}"
```
Comment (fgsch) [inline rules/REQUEST-910-IP-REPUTATION.conf]: indentation

File: rules/REQUEST-910-IP-REPUTATION.conf [hunk 2]
```diff
@@ -154,118 +154,120 @@ SecRule &TX:block_suspicious_ip "@eq 0" \
...
     tag:'attack-reputation-ip',\
-    chain,\
-    setvar:tx.httpbl_msg=%{tx.0}"
+    setvar:tx.httpbl_msg=%{tx.0},\
+    chain"
     SecRule TX:httpbl_msg "@rx RBL lookup of .*?.dnsbl.httpbl.org succeeded at TX:checkip. (.*?): .*" \
-        "t:none,\
-        capture,\
+        "capture,\
+	t:none,\
```
Comment (fgsch) [inline rules/REQUEST-910-IP-REPUTATION.conf]: indentation

File: rules/RESPONSE-980-CORRELATION.conf [hunk 3]
```diff
@@ -22,63 +22,66 @@
...
+    pass,\
     t:none,\
     log,\
-    pass,\
+    msg:'Correlated Successful Attack Identified: (Total Score: %{tx.anomaly_score}) Inbound Attack (%{tx.inbound_tx_msg} - Inbound Anomaly Score: %{TX.INBOUND_ANOMALY_SCORE}) + Outbound Data Leakage (%{tx.msg} - Outbound Anomaly Score: %{TX.OUTBOUND_ANOMALY_SCORE})',\
     tag:'event-correlation',\
-    skipAfter:END_CORRELATION,\
     severity:'EMERGENCY',\
-    msg:'Correlated Successful Attack Identified: (Total Score: %{tx.anomaly_score}) Inbound Attack (%{tx.inbound_tx_msg} - Inbound Anomaly Score: %{TX.INBOUND_ANOMALY_SCORE}) + Outbound Data Leakage (%{tx.msg} - Outbound Anomaly Score: %{TX.OUTBOUND_ANOMALY_SCORE})'"
-        SecRule &TX:'/WEB_ATTACK/' "@ge 1" "t:none"
```
Comment (fgsch) [inline rules/RESPONSE-980-CORRELATION.conf]: indentation

**Comment by dune73** [comment]
I have added expirevar to the list of actions in the initial comment of @fzipi on this page.

**Comment by dune73** [comment]
The merging of #838 lead to new conflicts. Sorry, but either this or conflicts in #838.

**Comment by fzipi** [comment]
Solved conflicts from #838 merge, and also bad indentation (thanks @fgsch !).

**Comment by dune73** [comment]
OK. Time to merge. Thank you for your work.

**Pull request merged**